llvm/include/llvm/Telemetry/Telemetry.h

```diff
@@ -0,0 +1,107 @@
+//===- llvm/Telemetry/Telemetry.h - Telemetry -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_TELEMETRY_TELEMETRY_H
+#define LLVM_TELEMETRY_TELEMETRY_H
+
+#include <chrono>
+#include <ctime>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
+
+namespace llvm {
+namespace telemetry {
+
+using SteadyTimePoint = std::chrono::time_point<std::chrono::steady_clock>;
+
+struct TelemetryConfig {
+  // If true, telemetry will be enabled.
+  bool enable_telemetry;
+
+  // Additional destinations to send the logged entries.
+  // Could be stdout, stderr, or some local paths.
+  // Note: these are destinations are __in addition to__ whatever the default
+  // destination(s) are, as implemented by vendors.
+  std::vector<std::string> additional_destinations;
+};
+
+struct TelemetryEventStats {
+  // REQUIRED: Start time of event
+  SteadyTimePoint m_start;
+  // OPTIONAL: End time of event - may be empty if not meaningful.
+  std::optional<SteadyTimePoint> m_end;
+  // TBD: could add some memory stats here too?
+
+  TelemetryEventStats() = default;
+  TelemetryEventStats(SteadyTimePoint start) : m_start(start) {}
+  TelemetryEventStats(SteadyTimePoint start, SteadyTimePoint end)
+      : m_start(start), m_end(end) {}
+
+  std::string ToString() const;
```

Comment:
Related to my above comment, this ties the writing of telemetry data to making it a string, which isn't really optimal. I think it would make more sense to have another class that handles writing of an event into whatever underlying storage is desired. For example, you might have an abstract base `Serializer` class that provides the interface, with concrete subclasses that provide `JsonSerializer` or `DictionarySerializer` for writing to a JSON file or a dictionary etc. Rather than have a `ToString` method, you'd have a `serialize` method which takes a `Serializer` and calls methods in its interface to write the data.

Reply:
I would call it a `Consumer` base class to avoid the connection to serialization.

Reply:
Not sure I see what's wrong with that? The destination and serialization are pretty tied together, at least in this kind of design. You could split them, in theory, but not all destinations will accept the same kind of serialization output, e.g. a file or stdout will take strings, whatever format they're in, but a dictionary class won't take arbitrary strings - it needs to know the types of the source data.

Reply:
It could be a `StatisticsConsumer`, which is only interested in the error rate.

Reply:
Actually, this `ToString` is intended to be mostly for human and/or descriptive purposes (sort of like th `Object::toString()`). It's not meant as a way to serialize data.

How the telemetry data get transmitted (or serialized) elsewhere is supposed to be entirely "hidden" within the implementation of a `Destination` class.

For eg., if a vendor wants to have the data serialized to json, sure they can define a Destination class that take a `TelemetryInfo` object and emit a json, etc.

Reply:
(as mentioned in the other comment, I've added a set of tests, which hopefully demonstrated how the pieces should interact. Please have a look. Thanks!)